### DISCUSSION TITLE
Best approach to know hydration is completed

### DISCUSSION THREAD

**Discussion by mdupree** [vercel/next.js#36689] [Help] [answered]
This is similiar in regards to https://github.com/vercel/next.js/discussions/18271

I am looking for the best approach for determining is hydration is complete. 

The Problem:
I have some code wthin `_app.jsx`  that needs to run once as it updates dataLayer and will cause issues with my analytics if it is loaded twice. 

Right now as I understand it if the page contains a query param then the call to dataLayer is run once prior to hydration and once after hydration.

The Solution I am using:
Create a `useEffect` with `next/router` as a dependency. Within the `useEffect` I am checking if `router.isReady` is `true`.

When it is true I run my sensitive function.

IE:
```javascript
  useEffect(() => {
    // Don't run until hydration is complete.
    if (router.isReady) {
      updateDataLayer(...);
    }
  }, [router.isReady]);
  ```

I am looking to know if this approach is reasonable, Or if there are better alternatives.

**Comment by icyJoseph** [comment, marked as answer]
Why wouldn't it be enough to place it in a useEffect with no dependencies?

Just to make sure we are talking about the same thing: https://github.com/vercel/next.js/discussions/36336#discussioncomment-2608125.

Remember than in the React 18 Strict Mode, in development, does a full mount/unmount cycle. So the empty dependencies useEffect, in development mode, apparently, runs twice (in reality it runs once every time it is mounted). 

In production it will run once, because, again, Strict Mode only happens in development. Which saves you the need to check if Next.js has executed this step:

> Pages that are statically optimized by [Automatic Static Optimization](https://nextjs.org/docs/advanced-features/automatic-static-optimization) will be hydrated without their route parameters provided, i.e query will be an empty object ({}).
> After hydration, Next.js will trigger an update to your application to provide the route parameters in the query object.

Since your `useEffect`, as shown in the example, has no dependency on the query param, then you don't really need to do any further checks.

**Reply by mdupree** [reply]
Yeah this is the same thing I am seeing, when a query param exists it being run twice (once without the query once with)

However this is happening in production. Though I am not running React v18, I'm currently running ^v17

When a user navigates to a url with a query parameter I am seeing the behaviour where the static page is rendered and then after hydration it is re-rendered with the query param. Which caused an issue where my updateDatalayer function would be called twice.

Though I am not dependant on the query param to run this function, I do need to know when hydration has completed so that this function runs only once, and only on the final render. (without a query param I dont have the issue with datalayer as it is only rendered once, however if a query param exists it renders twice and I need to call datalayer after hydration when the query parameter exists)

**Reply by mdupree** [reply]
I'm going to look at updating to react 18 as it sounds like that will solve this issue and I might not need to rely on checking `router.isReady`

**Reply by icyJoseph** [reply]
But how are you calling `updateDatalayer`, are you doing directly in the body of the function, or inside a `useEffect`?

`useEffect`, without dependencies, at the highest level (page or _app), runs once the DOM is React is done with its job.

Perhaps I am missing what's your actual requirement here, but `useEffect` should suffice. 

```tsx
// pages/_app.tsx
import type { AppProps } from "next/app";
import { useRouter } from "next/router";
import { useEffect } from "react";

function MyApp({ Component, pageProps }: AppProps) {
  const { query } = useRouter();

  console.count("_app Render");

  useEffect(() => {
    console.count("_app Mount");
  }, []);

  return <Component {...pageProps} />;
}

export default MyApp;
```

```tsx
// pages/index.tsx
import type { NextPage } from "next";
import { useRouter } from "next/router";
import { useEffect } from "react";

const Home: NextPage = () => {
  const { query } = useRouter();

  console.count("Home Render");
  console.log("Query:", query);

  useEffect(() => {
    console.count("Home Mount");
  }, []);

  return <div>Home</div>;
};

export default Home;
```

When build and started at `http://localhost:3000/?q=0`, logs:

```
_app Render: 1
Home Render: 1
Query: {}
Home Mount: 1
_app Mount: 1
_app Render: 2
Home Render: 2
Query: {q: '0'}
```

Perhaps you could describe a bit more what is it that you want to do?

**Reply by mdupree** [reply]
Yeah, I can elaborate. It's probably not very clear as I was intentionally leaving some data out that would be necessary for a full understanding.

Originally datalayer was being called like so, which worked fine if there was no query attached to the url.

```javascript
export default function MyApp({ Component, pageProps }) {
  const [resourcesInitialLoad, setResourcesInitialLoad] = useState(true);
  const pageTitle = getPageTitle(pageProps);

  useEffect(() => {
    // Initialize Google Tag Manager.
    initDataLayer();
  }, []);

  updateDataLayer(pageTitle, pageProps, resourcesInitialLoad, setResourcesInitialLoad);
  useSmoothScroll();
  
  return(...)
}
```

It was flagged to me by a developer that they were seeing datalayer object being updated twice.

My first instinct was to do as you suggested. move it into a useEffect hook.

```javascript
export default function MyApp({ Component, pageProps }) {
  const [resourcesInitialLoad, setResourcesInitialLoad] = useState(true);
  const pageTitle = getPageTitle(pageProps);

  useEffect(() => {
    // Initialize Google Tag Manager.
    initDataLayer();
  }, []);

  useEffect(() => {
    console.log('use effect', { pageTitle, pageProps });
    updateDataLayer(pageTitle, pageProps, resourcesInitialLoad, setResourcesInitialLoad);
  }, [pageProps, pageTitle, resourcesInitialLoad]);
  
  useSmoothScroll();
  
  return(...)
}
```

However this still runs twice if there is a query parameter attached to the url, due to pageProps updating.

Console logs
```
use effect 
{pageTitle: undefined, pageProps: {…}}

use effect 
{pageTitle: 'Example Page Title', pageProps: {…}}

```

(after first load, subsequent refreshes appears to work fine)

Console logs
```
use effect 
{pageTitle: 'Example Page Title', pageProps: {…}}

```

So then I did some digging and found that information you posted above
> Pages that are statically optimized by [Automatic Static Optimization](https://nextjs.org/docs/advanced-features/automatic-static-optimization) will be hydrated without their route parameters provided, i.e query will be an empty object ({}).
After hydration, Next.js will trigger an update to your application to provide the route parameters in the query object.

So that is what led me to needing to know when hydration was complete.
Which sounds like `Router.isReady` signifies this.

So I altered the useEffect like so

```javascript
export default function MyApp({ Component, pageProps }) {
  const router = useRouter();
  const [resourcesInitialLoad, setResourcesInitialLoad] = useState(true);
  const pageTitle = getPageTitle(pageProps);

  useEffect(() => {
    // Initialize Google Tag Manager.
    initDataLayer();
  }, []);

  useEffect(() => {

    // Don't run until hydration is complete.
    if (router.isReady) {
      console.log('use effect', { pageTitle, pageProps });
      updateDataLayer(pageTitle, pageProps, resourcesInitialLoad, setResourcesInitialLoad);
    }
  }, [pageProps, pageTitle, resourcesInitialLoad, router.isReady]);
  
  useSmoothScroll();
  
  return(...)
}
```

Which now seems to cover the issue. The call to datalayer only happens once hydration is completed. 
the function runs when it should with or without a query parameter.

Console logs
```
use effect 
{pageTitle: 'Example Page Title', pageProps: {…}}
```

So the code works. I just wasn't sure if there was perhaps a better way to handle this type of scenario.

**Reply by mdupree** [reply]
I should add, all pages are dynamic and built use static generation. The page data and routes comes from a backend CMS

**Reply by icyJoseph** [reply]
I see, well, first of all, the React version has very little to do here.

Every once in a blue moon, and when it is actually the requirement, you can skip the dependencies **if you know why, the consequences, and document the reason, to help your future self**. In this case for example the entire requirement is around just calling the function once. The body of the function happens on every render, and  useEffect with their dependencies specified, will run as soon as any changes. 

One solution could be to implement a flag (a closure) that is set when calling `updateDatalayer` the first time, so that the function internally knows to not execute further. You can even wrap this into a class, or keep it at module level. Or even better define a function `updateDatalayerOnce`.

```ts
let initialized = false;

// you'd pass the last argument as true to the `updateDatalayer` that's meant to be called on start.
const updateDatalayer = (/* arguments */, initial = false) => {

if (initial && initialized) return; // something along these lines

if (initial) {
 initialized = true;
}
// do the normal work
}
```

One can get really creative with the above, and also make it very complex, but it'd be totally independent of React's rules.

Another React solution could've been to put the arguments passed to `updateDatalayer` inside a `useRef`, and have an empty dependencies array in `useEffect`, you could also make the `useRef` contain a getter function to give you the arguments.

```ts
const layerArgs = useRef({pageTitle, pageProps, resourcesInitialLoad, setResourcesInitialLoad});

useEffect(() => updateDatalayer(layerArgs.current) ,[]);
```

You could also reach out for `useConstant` which is a simple hook you can implement yourself like:

```ts
// hooks/useConstant.tsx

const None = Symbol("none");

const useConstant = <Value>(fn: () => Value): Value => {
  const ref = useRef<Value | typeof None>(None);

  if (ref.current === None) {
    ref.current = fn();
  }

  return ref.current;
};
```

and then:

```ts
// right in the body of the function
useConstant(() => updateDatalyer(pageTitle, pageProps, resourcesInitialLoad, setResourcesInitialLoad))
```

Though maybe the last one, might suffer from concurrent features, I am not 100% sure.

**Reply by mdupree** [reply]
Awesome, I appreciate you taking the time to give me such a detailed response. This has given me some ideas for how I can improve the code a little bit. 

the main problem for me was just getting the function to run at the correct time (given there may or may not be a query param) and I need to function to run with the query parameter when it exists. 

I will consider my options here and do some testing.

**Reply by icyJoseph** [reply]
Nice! Don't forget about `asPath` from the `useRouter` hook, which will include the query from the first frame! So you don't have to think about whether the query update happens or not. And last but not least there's the `window.location.search` which natively has what you need, albeit, both of these have the query stringified, you can always build and object from that.
